### PR TITLE
Subatlas command now supports predicate slicing

### PR DESCRIPTION
**Pull request by lucaspcram**
### Description:

**DNM Notes**
~Please do not merge until the Groovy updates are in place.~
**READY FOR MERGE**

The `subatlas` command now supports slicing with predicates. For example, the following command would slice a subatlas containing only the road network in France:

```
$ atlas subatlas FRANCE.atlas --predicate 'e.getTag("highway").isPresent()'
```

You can also combine the `--predicate` option with the `--polygon` option - this will perform a predicate slice followed by a polygon slice. For example, we can cut a smaller, Parisian-road-network-only atlas from all of France with this command:

```
$ atlas subatlas FRANCE.atlas \
--predicate 'e.getTag("highway").isPresent()' \
--polygon 'POLYGON((1.8292419541808158 49.112967744380605,2.872943126055816 49.112967744380605,2.872943126055816 48.548870883220054,1.8292419541808158 48.548870883220054,1.8292419541808158 49.112967744380605))'
```

The argument to `--predicate` is a Groovy closure. By default, the provided closure is injected into a Groovy context that imports some useful default packages (like `org.openstreetmap.atlas.geography.atlas.items.*`) and converts the closure to a Java `Predicate<AtlasEntity>`, which is required by the underlying `subatlas` API.

Thanks to @yazad3 for all his help on this PR!

### Potential Impact:

This PR adds the Groovy language as a dependency to `atlas`. Impact should be minimal - the fat JARs remain roughly the same size.

~Security concerns: the injected Groovy closure is unvetted. This means that users can inject arbitrary code into the application. This is probably OK, provided this command is not running as a remote service and/or with elevated privileges.~
We are now doing some basic vetting using the `SecureASTCustomizer` as well as the `ImportCustomizer`. Thanks @yazad3 for helping me set this up!

### Unit Test Approach:

Tested on the command line using local atlas files. Also some unit tests included in `SubAtlasCommandTest`.

### Test Results:

Tests OK.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/resources/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommandScriptTemplate.groovy [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import java.util.function.*;
+import org.openstreetmap.atlas.geography.atlas.items.*;
+Predicate<AtlasEntity> predicate = {
+    %s;
+};
+
+return predicate;
```
Comment (yazad3) [inline src/main/resources/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommandScriptTemplate.groovy]: You may be able to do without any string manipulation.

Comment (lucaspcram) [reply src/main/resources/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommandScriptTemplate.groovy]: Ah yes this is a good idea. Let's not merge until implemented.

**Comment by yazad3** [comment]
@lucaspcram - the tests I added in this PR were mostly to share with you my suggestions, feel free to delete the tests if the code smells are an issue. I have approved the PR for the changes I requested anyways.

**Comment by lucaspcram** [comment]
@yazad3 I wanted to leave them just in case they catch some random issue in the future. However, it looks like Sonar is complaining that they don't have asserts. I'll update this soon so we can merge with no sonar failures.

**Comment by yazad3** [comment]
@lucaspcram - yes the stuff I added was more to be able to communicate what I was requesting, so feel free to take a call on deleting it. If however you need help in improving the test, let me know.

File: src/main/java/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommand.java [hunk 2]
```diff
@@ -148,19 +178,119 @@ protected void processAtlas(final Atlas atlas, final String atlasFileName,
...
+                this.outputDelegate.printlnErrorMessage("could not parse predicate");
+                return 1;
+            }
+            this.matcher = predicate;
+        }
+
         return 0;
     }
+
+    private Optional<Predicate<AtlasEntity>> getPredicateFromCommandLineExpression(
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommand.java]: Can this be broken out in a separate converter?

Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommand.java]: It might be useful in other contexts!

Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/utilities/command/subcommands/SubAtlasCommand.java]: This is a good idea.

File: src/main/java/org/openstreetmap/atlas/utilities/conversion/StringToPredicateConverter.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import groovy.lang.Binding;
+import groovy.lang.GroovyClassLoader;
+import groovy.lang.GroovyCodeSource;
+import groovy.lang.GroovyShell;
+import groovy.lang.Script;
+
+/**
+ * Convert a boolean expression string to a {@link Predicate}. While the converter can handle
+ * generic predicates, it includes some default imports that make it ideal for use as
+ * Predicate&lt;AtlasEntity&gt;. The converter uses the Groovy interpreter to create a
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/utilities/conversion/StringToPredicateConverter.java]: Awesome!

**Pull request merged**